src/node.cc

```diff
@@ -783,85 +781,115 @@ void RunBeforeExit(Environment* env) {
     EmitBeforeExit(env);
 }
 
-inline int StartNodeWithIsolate(Isolate* isolate,
-                                IsolateData* isolate_data,
-                                const std::vector<std::string>& args,
-                                const std::vector<std::string>& exec_args) {
+// TODO(joyeecheung): align this with the CreateEnvironment exposed in node.h
+// and the environment creation routine in workers somehow.
+inline std::unique_ptr<Environment> CreateMainEnvironment(
+    IsolateData* isolate_data,
+    const std::vector<std::string>& args,
+    const std::vector<std::string>& exec_args,
+    int* exit_code) {
+  Isolate* isolate = isolate_data->isolate();
   HandleScope handle_scope(isolate);
+
+  // TODO(addaleax): This should load a real per-Isolate option, currently
+  // this is still effectively per-process.
+  if (isolate_data->options()->track_heap_objects) {
```

Comment:
My thinking was when we implement snapshot, this could be where the context is deserialized from the snapshot (as opposed to `NewContext`). Note that to deserialize the snapshot properly we need to reinstall per-isolate callbacks, including the ones installed in `SetIsolateUpForNode`, so we may need to stop using `NewIsolate` at that point. We also need to restore inspector states after deserialization as those are discarded when the snapshot is generated, so it makes sense to me to encapsulate all this together.